src/tamp_improv/approaches/improvisational/graph_training.py

```diff
@@ -511,6 +531,131 @@ def identify_shortcut_candidates(
     return shortcut_candidates
 
 
+def identify_promising_shortcuts_with_rollouts(
+    system,
+    planning_graph,
+    observed_states: dict[int, Any],
+    num_rollouts_per_node: int = 50,
+    max_steps_per_rollout: int = 30,
+    shortcut_success_threshold: int = 1,
+    random_seed: int = 42,
+) -> list[ShortcutCandidate]:
+    """Identify promising shortcuts by performing random rollouts from each
+    node.
+
+    This function runs random rollouts from each source node in the planning graph
+    and tracks which preimages/nodes are reached during exploration. Shortcuts that
+    are reached at least `shortcut_success_threshold` times through random exploration
+    are considered promising candidates for training.
+    """
+    print("\n=== Identifying Promising Shortcuts with Random Rollouts ===")
+
+    np.random.seed(random_seed)
```

Comment:
let's just completely avoid using `np.random`. it's too confusing because it modifies global state. if some third party library that you're using also relies on `np.random` then things can get very confusing. use `rng = np.random.default_rng(seed)` instead